lib/core/src/Cardano/Wallet/StakePool/Metrics.hs

```diff
@@ -0,0 +1,67 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedLabels #-}
+-- | This module can fold over a blockchain to collect metrics about
+-- Stake pools.
+--
+-- It interacts with:
+-- - "Cardano.Wallet.Network" which provides the chain
+-- - "Cardano.Wallet.DB" - which can persist the metrics
+-- - "Cardano.Wallet.Api.Server" - which presents the results in an endpoint
+module Cardano.Wallet.StakePool.Metrics where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types
+    ( BlockHeader (..), PoolId (..), SlotId (..) )
+import Data.Map.Strict
+    ( Map )
+import Data.Word
+    ( Word64 )
+import Fmt
+    ( Buildable (..), blockListF', fmt, (+|), (|+) )
+import GHC.Generics
+    ( Generic )
+
+import qualified Data.Map.Strict as Map
+
+activityForEpoch :: Word64 -> State -> Map PoolId Int
+activityForEpoch epoch s =
+    Map.filter (> 0)
+    $ Map.map (length . filter slotInCurrentEpoch)
+    (activity s)
+  where
+    slotInCurrentEpoch = ((epoch ==) . epochNumber)
+
+--
+-- Internals
+--
+
+-- | In-memory state keeping track of which pool produced blocks at which slots.
+data State = State
+    { tip :: BlockHeader
+    , activity :: Map PoolId [SlotId]
```

Comment:
don't we want this to be `activity :: Map PoolId [(Word64,SlotId)]` ? ei., follow (epoch, slotId)? (this stop to be valid when we introduce absolute slotId soon)

Reply:
`tip` functions here as an identifier up till now we gathered statistics, right? The question is what is the start? I presume, the beginning of the current epoch..but what if we got crashed in the middle... Why not have this instead : `[[BlockHeader]]` to exactly state what intervals the statistics cover?

Reply:
> The question is what is the start? I presume, the beginning of the current epoch

It has to start at genesis (we can't reliably know the `BlockId` of the start of the recent epoch). Also… I'm now thinking we should skip/delay clearing everything but the most recent epoch, since it is non-trivial to implement (with rollbacks especially, I think), and we might have access to stake distributions for all epochs.

> but what if we got crashed in the middle

This model isn't meant to be persisted (will start from the beginning again if crashed).

Reply:
ok, so `tip` means - statistics were taken from the beginning to the block identified by block header.

Reply:
> (this stop to be valid when we introduce absolute slotId soon)

That was part of the reason for https://github.com/input-output-hk/cardano-wallet/pull/753 😁 So we should hopefully be able to use the original `(epoch, slotNo)` from the blocks here, even if the walletLayer changes. Also, I think #729 might be on hold.